src/optimize.rs

```diff
@@ -3,298 +3,493 @@ use egg::*;
-    let mut g = egraph.analysis.graph.borrow_mut();
-    match enode {
-        Mdl::Num(_)
-        | Mdl::Var(_)
-        | Mdl::Input(_)
-        | Mdl::Weight(_)
-        | Mdl::Merge(_)
-        | Mdl::Split0(_)
-        | Mdl::Split1(_)
-        | Mdl::Reshape(_)
-        | Mdl::Transpose(_) => 0.0,
-
-        Mdl::Relu(_a) => {
-            // Check types
-            let a_t_data = x(_a);
-            assert!(a_t_data.dtype == DataKind::Tnsr);
-
-            unsafe {
-                // Get op
-                let op = (*g.model).get_or_create_activation(*a_t_data.meta, OpType_OP_RELU, true);
-                assert!(op != Op_INVALID_OP);
-                (*op.ptr).runtime.clone()
-            }
+/// Class for our cost model
+pub struct CostModel {
+    /// To have zero cost for all weight op only
+    ignore_all_weight_only: bool,
+    /// Discount factor for all weight ops
+    all_weight_discount: f32,
+}
+
+impl CostModel {
+    pub fn with_setting(ignore_all_weight_only: bool) -> Self {
+        CostModel {
+            ignore_all_weight_only: ignore_all_weight_only,
+            all_weight_discount: 1.0,
         }
+    }
 
-        Mdl::Tanh(_a) => {
-            // Check types
-            let a_t_data = x(_a);
-            assert!(a_t_data.dtype == DataKind::Tnsr);
+    /// Gets cost for the enode itself.
+    ///
+    /// This function gets the cost by calling TASO's get_or_create_{some_op}()
+    /// functions with the tensor information stored in metadata. TASO side stores
+    /// hashmaps for OpBase objects. So here TASO side will simply lookup previously
+    /// created ops (with previously measured runtime).
+    ///
+    /// # Parameters
+    ///
+    /// - `egraph`: E-graph of interest
+    /// - `enode`: enode to get cost for
+    ///
+    /// # Returns
+    ///
+    /// Cost for this enode.
+    pub fn get_self_cost(&self, egraph: &EGraph<Mdl, TensorAnalysis>, enode: &Mdl) -> f32 {
+        let x = |i: &Id| &egraph[*i].data;
+        let mut g = egraph.analysis.graph.borrow_mut();
+        match enode {
+            Mdl::Num(_)
+            | Mdl::Var(_)
+            | Mdl::Input(_)
+            | Mdl::Weight(_)
+            | Mdl::Merge(_)
+            | Mdl::Split0(_)
+            | Mdl::Split1(_)
+            | Mdl::Reshape(_)
+            | Mdl::Transpose(_)
+            | Mdl::Noop(_) => 0.0,
+
+            Mdl::Relu(_a) => {
+                // Check types
+                let a_t_data = x(_a);
+                assert!(a_t_data.dtype == DataKind::Tnsr);
+
+                let runtime = unsafe {
+                    // Get op
+                    let op =
+                        (*g.model).get_or_create_activation(*a_t_data.meta, OpType_OP_RELU, true);
+                    assert!(op != Op_INVALID_OP);
+                    (*op.ptr).runtime.clone()
+                };
+
+                if self.ignore_all_weight_only && x(_a).all_weights {
+                    self.all_weight_discount * runtime
```

Comment:
What is the value of self.all_weight_discount? I see that the default is 1. If it is 1, then this does nothing. Who set this value?

Reply:
Currently it's just 1.0. We can add a command line flag for it if we decide to experiment with this